sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+```
+
+
+## Detailed Design
+
+### Managing the Graph and Ops Accessors
+To use TensorFlow, we must have access to a `Graph` to build and run computations. In Python, this is constructed implicitly. In Java, the `Graph` and `Ops` objects must be created explicitly. Thus, we
+leave this (below) for the user to write, and allow the `Ops` object to be passed
+throughout keras `Layer` construction calls to provide access to core TensorFlow operations.
+
+```java
+try(Graph graph = new Graph) {
+    Ops tf = Ops.create(graph);
+    
+    // Keras code here.
+}
+```
+
+When using the Keras API, models are "compiled" and "fitted" within this block, and the `tf` object is passed through the internal method calls.
+
+### Keras Identifiers
+Standard Keras objects are very often referenced just by name, and don't all have
+to be created explicitly by the user.
+For example, in Python, the strings `"sgd"` or `"adam"` can be used to create different
+kinds of `Optimizer`s, `"relu"` and `"softmax"` stand for specific `Activations`, etc.
+
+In Java, the supported standard types are enumerated in various enum classes: `Activations`, 
+`Intializers`, `Optimizers`, etc., so these types become `Optimizers.sgd`, `Activations.relu`,
+etc. Using enums instead of strings seems to result in greater visibility into all possible
+types, and reduces error-proneness.
+
+Each of these classes has a `select` method which convers the enum value to an object of the
+target type. For example, `Optimizers.select(Optimizers.sgd)` returns an `Optimizer`. However
+the end users shouldn't need to interact with this.
+
+### Keras Configuration
+Currently, the `~/.keras` directory is used for storing downloaded files, and user configurations 
+(for now, just datafiles for standard datasets). This works well when the Java keras client is used
+in isolation, but potentially could cause conflicts when a user is (likely) using the Python client
+for other projects.
+
+### Standard Datasets
+A few standard datasets are defined in `org.tensorflow.keras.datasets` (currently,
+FashionMNIST and MNIST). These classes download datasets in their original format,
+and define batched loaders for loading tensors. 
+
+
+### Options and Builder Classes
+
+To enable methods and constructors to use optional arguments with defaults,
+`Options` classes are defined (most notably for `Layer` construction,
+and `compile` and `fit` configuration for models). The structure for an `Options` class
+of object `KerasObject` is standardized as follows; if changes can be made to this structure
+so that it is cleaner and easier to work with, please comment!
+
+```java
+class KerasObject {
+
+    static {
+        Arg1 arg1;
+        Arg2 arg2;
+        
+        // Create KerasObject from Options class
+        KerasObject kerasObject =
+            new KerasObject(KerasObject.Options.builder()           
+                .setArg1(arg1)
+                .setArg2(arg2)
+                .build());
+    }
```

Comment:
Not sure to understand this code... where is initialized `arg1` and `arg2`? and what the `kerasObject` is used for?

Reply:
This is just an example structure for any Object we want to define that takes multiple optional/default parameters; hoping to standardize this structure across `Layer`, `Optimizer`, `Initializer`... arg1 and arg2 are null here until initialized via a builder call to `setArg1` or `setArg2`